File: scala/src/main/scala/Triangle.scala
class Triangle(var l1:Double, var l2:Double, var l3:Double) {
  def legal: Boolean = {
    var legal = l1 > 0 && l2 > 0 && l3 > 0
    if (legal) {
      val longest = List.apply(l1, l2, l3).max
      legal = longest < (l1 + l2 + l3) - longest
    }
    legal    
  }
  
  def equilateral: Boolean = {
    legal && l1 == l2 && l2 == l3    
  }

  def isosceles: Boolean = {
    legal && l1 == l2 || l1 == l3 || l2 == l3     
  }

  def scalene: Boolean = {
    legal && l1 != l2 && l1 != l3 && l2 != l3    
  }
}

object Triangle {
    def apply(l1:Double, l2:Double, l3:Double): Triangle = {
        new Triangle(l1, l2, l3)
    }
}

File: python/src/triangle.py
def triangle(sides):
    result = True
    if len(sides) != 3 or sides[0] <= 0 or sides[1] <= 0 or sides[2] <= 0:
        result = False
    else:
        sides.sort()
        result = sides[0] + sides[1] > sides[2]

    return result

def equilateral(sides):
    return triangle(sides) and len(set(sides)) == 1


def isosceles(sides):
    return triangle(sides) and len(set(sides)) <= 2


def scalene(sides):
    return triangle(sides) and len(set(sides)) == 3
